streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -553,7 +553,7 @@ private void handleTasksWithStateUpdater(final Map<TaskId, Set<TopicPartition>>
     private void handleTasksPendingInitialization() {
         // All tasks pending initialization are not part of the usual bookkeeping
         for (final Task task : tasks.drainPendingTasksToInit()) {
-            closeTaskClean(task, Collections.emptySet(), Collections.emptyMap());
+            closeTaskClean(task, new HashSet<>(), new HashMap<>());
```

Comment:
Is it necessary to call `closeDirty` for those failed tasks?

Reply:
Good point. I guess yes. (even if it seems unlikely that a task that was initialized yet, would fail to close clean).

Originally we did not use `closetTaskClean()` but the PR introducing this change started to use it: https://github.com/apache/kafka/pull/16730

I did not did into the original PR, but it seems there was some issue with leaking resources in tests? So maybe it's just a test issue?

Reply:
I am frankly not sure, why https://github.com/apache/kafka/pull/16730 would actually address any resource leak, and the old and new code would just call
```
            task.suspend();
            task.closeClean();
```

The difference is just, that the new code would catch and swallow some exception.

So maybe if we throw some exception, other cleanup code did not run leaking the producer? But with the new code, we would swallow the exception ensuring that other cleanup code does run?

Would be good to understand better. Mabye @FrankYang0529 can shed some light?

Reply:
It's actually not even clear to me, how a "pending task to init" could fail to close clean to begin with... \cc @cadonna @lucasbru @ableegoldman and ideas?

Reply:
The leak resource was found by https://github.com/apache/kafka/pull/14783. On that branch, the leak resource only happened for the case `EosIntegrationTest#shouldCheckpointRestoredOffsetsWhenClosingCleanDuringRestoringStateUpdaterEnabled(eosConfig=StreamsConfig.EXACTLY_ONCE, processingThreadsEnabled="false")`. I think it's safe to revert the change cause of:

1. The EOSv1 was removed in 4.0. 
2. ClusterTestExtension also has [thread leak detection](https://github.com/apache/kafka/blob/810beef50e189cc9d1a0839961dffd14930ffe2e/test-common/test-common-runtime/src/main/java/org/apache/kafka/common/test/junit/ClusterTestExtensions.java#L164-L175). Revert the change of #16730 on the trunk branch, the test case still can pass.

Reply:
@FrankYang0529 thanks for the explanation. 

> how a "pending task to init" could fail to close clean to begin with... 

@mjsax yes, that is a good question ...

Reply:
Thanks @FrankYang0529 -- still not sure right now if just reverting is the right thing:

> that is a good question ...

The is the thing that worries me -- we did see it crash with EOSv2, so some pending tasks did get revoked and there was an error during revocation. So maybe we did expose to different bug.

Reply:
I was looking into the logs we collected more, and it seems possible that we get an error on close with EOSv2.

When we close the unitized tasks, we close the `RecordCollector` and we check for pending errors: https://github.com/apache/kafka/blob/4.0/streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java#L568

Because we shared a single producer across multiple task, the initialized task could observe an error from some other task. -- So it seems to be actually safe to swallow this exception. It's just not totally clear to me, how/where to implement this swallowing correctly?

Reply:
I believe something like in the following code, will avoid the leak:
https://github.com/apache/kafka/blob/efd785274e7c6dd75a3286ca6da1379830366acc/streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java#L1325
Basically, the task is closed and the exception is swallowed.

Reply:
@cadonna 's suggestion makes sense to me

Reply:
Thanks, agree. It was also what @chia7712 original comment was about:
> Is it necessary to call closeDirty for those failed tasks?

It was originally just not clear to me, why a uninitilzied task _could_ fail.